cloud, sql/sem/tree, backupccl, *: sanitize URLs during Format

michae2 opened the pull request:
Prior to this change, all SQL statements containing URLs would be formatted with the full URL in cleartext, including any secrets such as keys or passwords. These secrets would sometimes show up in the slow query log or sql audit log. This PR adds new AST nodes tree.URI and tree.URIs which are designed to sanitize their contents during formatting.

See individual commits for details.

Fixes: CRDB-39710, TREQ-284

Epic: None

Release note: None

michae2 committed:
cloud: move uris.go into separate cloud/uris package

This allows us to call `SanitizeExternalStorageURI` and `RedactKMSURI`
from sql/sem/tree without creating a circular dependency with cloud.

This was mostly mechanical, the only interesting part was that I had to
move `RegisterRedactedParams` into uris.go to make
`SanitizeExternalStorageURI` work correctly without a circular
dependency between cloud and cloud/uris.

Epic: None

Release note: None

michae2 committed:
tree, parser: add tree.URI and tree.URIs AST nodes

Add new AST nodes `tree.URI` and `tree.URIs`, which represent one or
more string literals containing URLs. During formatting the URLs are
sanitized to remove sensitive information such as passwords and keys.

There are many special cases in `tree.URI.Format` for which we do *not*
sanitize the URL:
- if the URL is a placeholder like $1
- if the URL is an empty string or an underscore
- if the URL has already been sanitized (this is done while building the
  output of various SHOW commands)
- if we're showing passwords, or already hiding constants

Future commits will use these AST nodes in all SQL statements that could
use a URL.

Epic: None

Release note: None

michae2 committed:
tree, parser, backupccl: use tree.URI in BACKUP and RESTORE

Use the `tree.URI` AST node in various backup-related statements.

Epic: None

Release note (security update): URLs in the following SQL statements are
now sanitized of any secrets before being logged (even to unredacted
logs):
- `BACKUP`
- `RESTORE`
- `CREATE SCHEDULE FOR BACKUP`
- `ALTER BACKUP SCHEDULE`

michae2 committed:
tree, parser, backupccl: use tree.URI in ALTER BACKUP

Use the `tree.URI` AST node in `ALTER BACKUP`.

Epic: None

Release note (security update): URLs in the following SQL statements are
now sanitized of any secrets before being logged (even to unredacted
logs):
- `ALTER BACKUP`

michae2 committed:
tree, parser, backupccl: use tree.URI in SHOW BACKUP

Use the `tree.URI` AST node in `SHOW BACKUP`.

Epic: None

Release note (security update): URLs in the following SQL statements are
now sanitized of any secrets before being logged (even to unredacted
logs):
- `SHOW BACKUP`
- `SHOW BACKUPS`

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/126185)
<!-- Reviewable:end -->

michae2 committed:
tree, parser, changefeedccl: use tree.URI in CREATE CHANGEFEED

Use the `tree.URI` AST node in `CREATE CHANGEFEED`.

Epic: None

Release note (security update): URLs in the following SQL statements are
now sanitized of any secrets before being logged (even to unredacted
logs):
- `CREATE CHANGEFEED`
- `CREATE SCHEDULE FOR CHANGEFEED`

michae2 committed:
tree, parser, importer: use tree.URI in IMPORT

Use the `tree.URI` AST node in `IMPORT`.

Epic: None

Release note (security update): URLs in the following SQL statements are
now sanitized of any secrets before being logged (even to unredacted
logs):
- `IMPORT`

michae2 committed:
tree, parser, opt: use tree.URI in EXPORT

Use the `tree.URI` AST node in `EXPORT`.

Epic: None

Release note (security update): URLs in the following SQL statements are
now sanitized of any secrets before being logged (even to unredacted
logs):
- `EXPORT`

michae2 committed:
tree, parser, externalconn: use tree.URI in CREATE EXTERNAL CONNECTION

Use the `tree.URI` AST node in `CREATE EXTERNAL CONNECTION`.

Note that this changes the output of `SHOW CREATE EXTERNAL CONNECTION`.

Epic: None

Release note (security update): URLs in the following SQL statements are
now sanitized of any secrets before being logged (even to unredacted
logs):
- `CREATE EXTERNAL CONNECTION`

michae2 committed:
tree, parser: use tree.URI in COPY

Use the `tree.URI` AST node in `COPY`.

Epic: None

Release note (security update): URLs in the following SQL statements are
now sanitized of any secrets before being logged (even to unredacted
logs):
- `COPY`

michae2 committed:
sql: remove tree.StringOrPlaceholderOptList

Epic: None

Release note: None

michae2 committed:
cloud/uris: break dependency on cloud/cloudpb

We need to break the dependency of cloud/uris on cloud/cloudpb to
satisfy various disallowed import constraints in sql/sem/tree when we
import cloud/uris from there. Fortunately this is straightforward: we
just have to move a few functions from uris.go to external_storage.go.

Epic: None

Release note: None

michae2 commented:
This is a big PR, so I'm hoping we can divide reviewing in the following way (1-based indexing):
- [ ] commit 1, 2, breaking cloud package dependencies: someone from DR, maybe @msbutler
- [ ] commit 3, adding tree.URI: someone from foundations or queries, maybe @rafiss and/or @mgartner
- [ ] commit 4, 5, 6, modifying BACKUP AND RESTORE: someone from DR, maybe @msbutler
- [ ] commit 7, modifying CREATE CHANGEFEED: someone from CDC, maybe @wenyihu6
- [ ] commit 8, 9, modifying IMPORT and EXPORT: someone from queries, @mgartner
- [ ] commit 10, modifying CREATE EXTERNAL CONNECTION: someone from DR, @msbutler
- [ ] commit 11, modifying COPY: someone from queries, @mgartner
- [ ] commit 12, removing an AST node: someone from queries or foundations, @mgartner or @rafiss

(Thanks in advance everyone, if you want to designate someone else to review in your place, that is fine.)

Additionally there's a need to backport this (or something equivalent) (see [this thread](https://cockroachlabs.slack.com/archives/C01C1HGC46Q/p1719844942840469) for context), so I'm hoping @mgartner and @rafiss could consider whether this PR could be backported as-is, or if we would need to de-risk this somehow.

michae2 commented on pkg/sql/sem/tree/uri.go:
```diff
@@ -0,0 +1,128 @@
+		} else {
+			sanitized, err = uris.SanitizeExternalStorageURI(raw, nil)
+		}
+		if err != nil {
```
There hasn't been a thorough discussion yet, so we can have it now.

The sanitization performed by `SanitizeExternalStorageURI` is already used by [backup](https://github.com/cockroachdb/cockroach/blob/d41cb10e932e9ecf9b7ac78336a6f81651ff3f76/pkg/ccl/backupccl/backup_planning.go#L159), [changefeeds](https://github.com/cockroachdb/cockroach/blob/d41cb10e932e9ecf9b7ac78336a6f81651ff3f76/pkg/ccl/changefeedccl/changefeed.go#L111), [create external connection](https://github.com/cockroachdb/cockroach/blob/d41cb10e932e9ecf9b7ac78336a6f81651ff3f76/pkg/sql/create_external_connection.go#L167), and [import](https://github.com/cockroachdb/cockroach/blob/d41cb10e932e9ecf9b7ac78336a6f81651ff3f76/pkg/sql/importer/import_planning.go#L238), so it seemed natural to me to also use it here. But it did require the first two commits (breaking circular dependencies). This seemed reasonable to me, since the compiler could help me perform the first two commits mechanically and confirm their correctness, but I understand that it creates a lot of code churn.

Pros:
- logged sanitized URL matches sanitized URL in SHOW output, job details, error messages, other places
- logged sanitized URL keeps non-sensitive information intact (domain, path, etc)
- same URL sanitization routine used everywhere, can just change that one routine

Cons:
- first two commits create code churn
- backport more difficult, though I would argue not necessarily more risky (if it compiles after the first two commits, they should be correct)

WDYT?

dt replied:
The granular per-param sanitization is performed in backup and friends since we thought it was necessary -- when showing you jobs, we think we need to show you where the job is backing up to, or what options storage options it is using when those are not sensitive. But in a generic log of all queries/slow queries/etc, I wonder if that still holds, or if all we really need to indicate is that it is a backup statement (and you can go consult the jobs if you want to see the details of the backup)?  

> same URL sanitization routine used everywhere, can just change that one routine

FWIW, the alternative I'm considering still only has one routine for _sanitizing_ a URI; I'd just be eliding it completely if I don't use that routine rather than having a distinct routine.

I'm not strongly opposed to the first two commits; just trying to think about the "change as little code as possible" backport guideline and if we really need sophisticated granular sanitization of URIs here at all, or could just elide them.

michae2 replied:
> I'm not strongly opposed to the first two commits; just trying to think about the "change as little code as possible" backport guideline and if we really need sophisticated granular sanitization of URIs here at all, or could just elide them.

Good point. @rytaft and @mgartner have also asked if I can figure out a simpler version for backport that elides the URLs completely. I will try that out, and then rebase this work on top of that simpler backportable change.

michae2 commented:
Marking this as a draft while I figure out a simpler story around backport. Hold off on reviewing for a bit.

dt commented:
Perhaps a smaller, more backport friendly change would be to leave these as strings/Expr but just extend Format() for those nodes that _contain_ them to elide rather than print based on the flag? I think that could get a similar result with only a handful of diffs in pkg/sql/sem/tree?

michae2 commented:
> Perhaps a smaller, more backport friendly change would be to leave these as strings/Expr but just extend Format() for those nodes that _contain_ them to elide rather than print based on the flag? I think that could get a similar result with only a handful of diffs in pkg/sql/sem/tree?

I think you're right. I started this last night: https://github.com/cockroachdb/cockroach/pull/126970. So far it seems like it will affect the output of some commands such as ALTER BACKUP and also telemetry. But I think it will work.

I'd like to bring back the calls to `SanitizeExternalStorageURI` on top of the simpler backportable change, because then we get most of the URLs in most places.

michae2 commented:
Closing this in favor of https://github.com/cockroachdb/cockroach/pull/126970 plus a future PR to add the call to `SanitizeExternalStorageURI`.
